signer/universal/universal_test.go

```diff
@@ -38,5 +287,345 @@ func TestNewSigner(t *testing.T) {
+	for name, profile := range universalConfig.Signing.Profiles {
+		if profile.RemoteServer != "" {
+			universalConfig.Signing.Profiles[name].RemoteServer = remoteServer.URL[7:]
+		}
+	}
+	s := newTestUniversalSigner(t, universalConfig.Signing)
+
+	for name, profile := range universalConfig.Signing.Profiles {
+		hosts := []string{"cloudflare.com"}
+		for _, test := range csrTests {
+			csr, err := ioutil.ReadFile(test.file)
+			if err != nil {
+				t.Fatalf("CSR loading error (%s): %v", name, err)
+			}
+			testSerial := big.NewInt(0x7007F)
+
+			certBytes, err := s.Sign(signer.SignRequest{
+				Hosts:   hosts,
+				Request: string(csr),
+				Serial:  testSerial,
+				Profile: name,
+			})
+			if test.errorCallback != nil {
+				test.errorCallback(t, err)
+			} else {
+				if err != nil {
+					t.Fatalf("Expected no error. Got %s. Param %s %d", err.Error(), test.keyAlgo, test.keyLen)
+				}
+				cert, err := helpers.ParseCertificatePEM(certBytes)
+				if err != nil {
+					t.Fatal("Fail to parse returned certificate:", err)
+				}
+				ku, _, _ := profile.Usages()
+				if cert.KeyUsage != ku {
+					t.Fatalf("Key usage was incorrect expected %+v, got %+v", ku, cert.KeyUsage)
+				}
+			}
+		}
+	}
+}
+
+func newConfig(t *testing.T, configBytes []byte) *config.Config {
+	conf, err := config.LoadConfig([]byte(configBytes))
+	if err != nil {
+		t.Fatal("config loading error:", err)
+	}
+	if !conf.Valid() {
+		t.Fatal("config is not valid")
+	}
+	return conf
+}
+
+func newTestUniversalSigner(t *testing.T, policy *config.Signing) signer.Signer {
+	h := map[string]string{
+		"key-file":  testCaKeyFile,
+		"cert-file": testCaFile,
+	}
+
+	r := &Root{
+		Config:      h,
+		ForceRemote: false,
+	}
+
+	s, err := NewSigner(*r, policy)
+	if err != nil {
+		t.Fatal("fail to init universal signer:", err)
+	}
+
+	return s
+}
+
+func newTestSignHandler(t *testing.T, s signer.Signer) (h http.Handler) {
+	h, err := NewSignHandlerFromSigner(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func newTestInfoHandler(t *testing.T, s signer.Signer) (h http.Handler) {
+	h, err := apiinfo.NewHandler(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func newTestSignServer(t *testing.T, s signer.Signer) *httptest.Server {
+	mux := http.NewServeMux()
+	mux.Handle("/api/v1/cfssl/sign", newTestSignHandler(t, s))
+	ts := httptest.NewUnstartedServer(mux)
+	ts.Start()
+	t.Log(ts.URL)
+	return ts
+}
+
+func newTestInfoServer(t *testing.T, s signer.Signer) *httptest.Server {
+	mux := http.NewServeMux()
+	mux.Handle("/api/v1/cfssl/info", newTestInfoHandler(t, s))
+	ts := httptest.NewUnstartedServer(mux)
+	ts.Start()
+	t.Log(ts.URL)
+	return ts
+}
+
+func closeTestServer(t *testing.T, ts *httptest.Server) {
+	t.Log("Finalizing test server.")
+	ts.Close()
+}
+
+// NewSignHandlerFromSigner generates a new SignHandler directly from
+// an existing signer.
+func NewSignHandlerFromSigner(s signer.Signer) (h http.Handler, err error) {
```

Comment:
woop sorry meant to do that, there was a point where i had changed it then
forgot to move back

On Thu, Oct 29, 2015 at 10:15 AM, Nick Sullivan <notifications@github.com>
wrote:

> In signer/universal/universal_test.go
> <https://github.com/cloudflare/cfssl/pull/399#discussion_r43415240>:
>
> > +	mux := http.NewServeMux()
> > +	mux.Handle("/api/v1/cfssl/info", newTestInfoHandler(t, s))
> > +	ts := httptest.NewUnstartedServer(mux)
> > +	ts.Start()
> > +	t.Log(ts.URL)
> > +	return ts
> > +}
> > +
> > +func closeTestServer(t *testing.T, ts *httptest.Server) {
> > +	t.Log("Finalizing test server.")
> > +	ts.Close()
> > +}
> > +
> > +// NewSignHandlerFromSigner generates a new SignHandler directly from
> > +// an existing signer.
> > +func NewSignHandlerFromSigner(s signer.Signer) (h http.Handler, err error) {
>
> This is duplicated from signer/remote/remote_test.go, it could be factored
> into.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cloudflare/cfssl/pull/399/files#r43415240>.
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>